pkg/internal/export/otel/metrics.go

```diff
@@ -231,6 +215,30 @@ func (mr *MetricsReporter) newMetricSet(service svc.ID) (*Metrics, error) {
 	return &m, nil
 }
 
+// evictTracer invokes the eviction callback for the cached Metrics. When the MetricsReporter
+// is running, the eviction callback is invoked asynchronously to avoid that any delay in the
+// remote flush affects the metrics processing.
+// When the MetricsReporter is closed, the eviction callback is evicted synchronously to make
+// sure that all the metrics are flushed before closing the exporters.
+func (mr *MetricsReporter) evictMetrics(id svc.ID, v *Metrics) {
+	llog := mlog().With("service", id)
+	llog.Debug("evicting metrics reporter from cache")
+	if mr.closed {
+		mr.closeProvider(llog, v.provider)
+	} else {
+		go mr.closeProvider(llog, v.provider)
+	}
+}
+
+func (mr *MetricsReporter) closeProvider(log *slog.Logger, tp *metric.MeterProvider) {
+	if err := tp.ForceFlush(mr.ctx); err != nil {
+		log.Warn("error flushing meter provider", "error", err)
+	}
+	if err := tp.Shutdown(mr.ctx); err != nil {
```

Comment:
I don't think this will help if I'm right about the issue. The docs on Shutdown says that it does flush, so this change I think it's going to bring back the old behaviour.  From the docs on shutdown:

```
// Shutdown shuts down the MeterProvider flushing all pending telemetry and
// releasing any held computational resources.
//
// This call is idempotent. The first call will perform all flush and
// releasing operations. Subsequent calls will perform no action and will
// return an error stating this.
//
```

I think the problem is to do with batching in our event processing. There will always be some delay between when we evict the exporter from the cache and when it's not used anymore. I suspect someone still uses the exporter and not all events are processed after the eviction calls shutdown, otherwise I can't explain this error in the customer logs:

```
time=2023-12-20T21:04:32.486Z level=DEBUG msg="evicting metrics reporter from cache" component=otel.TracesReporter service="{Name:livenessprobe Namespace: SDKLanguage:go Instance:beyla-n4h9q-3925388}"
2023/12/20 21:04:32 failed to upload metrics: gRPC exporter is shutdown
```

It must be some internal temporality in the OTel SDK, since our metric generation loop goes event by event and what we evicted is not used in the current loop iteration.